File: DataStructures/Stacks/leetcode/85. Maximal Rectangle/GPT/Maximal_Rectangle.py
# ### 1. 問題分析結果

# #### 競技プログラミング視点

# * **最速手法**: 各行を「高さ配列（連続1の本数）」に更新し、行ごとに **単調増加スタック**で「ヒストグラムの最大長方形」を計算。
#   計算量 **O(R·C)**、追加メモリ **O(C)**。200×200 の制約に十分高速。
# * **メモリ最小化**: `heights` と `stack` を固定長 1 本ずつ再利用。センチネルはループの境界で擬似化し、配列拡張を避ける。

# #### 業務開発視点

# * **可読性/保守性**: ヒストグラム計算をローカルメソッドに分離。変数名は意味ベースで統一。
# * **型安全性/エラーハンドリング**: 競プロ向け `maximalRectangle` は前提どおりの入力を想定（LeetCode仕様）。併せて堅牢版 `maximalRectangle_production` を用意し、行長不揃い・型不正時に例外を明示。

# #### Python特有分析

# * **CPython最適化**:

#   * ローカル変数バインドで属性参照を削減（`heights`, `stack`, `cols` をローカルに保持）。
#   * `list.append/pop` のみ使用、`deque`/`insert`/`pop(0)` は未使用。
#   * ループは `for` 直書き、内包表記は不要な中間生成を回避。
# * **GIL**: CPUバウンドの単一スレッド処理で問題なし。

# ---

# ### 2. アルゴリズム比較表

# | アプローチ                         | 時間計算量           | 空間計算量  | Python実装コスト | 可読性 | 標準ライブラリ活用   | CPython最適化 | 備考      |
# | ----------------------------- | --------------- | ------ | ----------- | --- | ----------- | ---------- | ------- |
# | 方法A: 行ごとヒストグラム + 単調増加スタック（採用） | O(R·C)          | O(C)   | 低           | 高   | list（固定再利用） | 適          | 最速・実装短  |
# | 方法B: DPで左右境界・高さを保持            | O(R·C)          | O(R·C) | 中           | 中   | なし          | 適          | メモリ重め   |
# | 方法C: 全探索（各セルを起点に拡張）           | O(R·C·min(R,C)) | O(1)   | 低           | 中   | なし          | 不適         | 小規模のみ妥当 |

# ---

# ### 3. Python最適化ポイント（抜粋）

# * **ローカル化**: `stack_top` を整数で手動管理し、`stack[-1]` 参照頻度を削減。
# * **センチネル**: 末尾 `j == cols` を高さ 0 として処理し、配列連結不要。
# * **再利用**: `heights` と `stack` は都度 `clear` せずインデックスで管理。

# ---

# ### 4. 実装（LeetCode形式 / Class / CPython 3.11+ / 型付き）

# ```python
from typing import List


class Solution:
    """
    LeetCode 85. Maximal Rectangle

    競技プログラミング最適化版: maximalRectangle
    - 前提: matrix は '0'/'1' の文字で構成された長方形2次元配列
    - 時間計算量: O(R*C)
    - 空間計算量: O(C)
    """

    def maximalRectangle(self, matrix: List[List[str]]) -> int:
        """
        与えられた '0'/'1' 行列で、1のみから成る最大長方形の面積を返す。

        Args:
            matrix: R x C の2次元配列（各要素は '0' または '1'）

        Returns:
            最大長方形の面積（int）

        Note:
            LeetCodeの想定に合わせ、入力検証は最小限（競技向け）。
        """
        if not matrix:
            return 0
        rows: int = len(matrix)
        cols: int = len(matrix[0])
        if cols == 0:
            return 0

        # heights[j]: 現在行を底とする列 j の連続 '1' の本数
        heights: List[int] = [0] * cols
        # 単調増加スタック（インデックスを格納）。top を手動管理。
        stack: List[int] = [0] * (cols + 1)  # 十分な容量を確保
        max_area: int = 0

        # 局所関数: ヒストグラムの最大長方形
        def largest_rectangle_in_histogram(h: List[int]) -> int:
            best: int = 0
            top: int = -1  # 空を -1 とする
            # j == cols のとき高さ0のセンチネルとして処理
            for j in range(cols + 1):
                cur: int = 0 if j == cols else h[j]
                # 単調性が崩れたら確定計算
                while top >= 0 and cur < h[stack[top]]:
                    height: int = h[stack[top]]
                    top -= 1
                    left_less_index: int = stack[top] if top >= 0 else -1
                    width: int = j - left_less_index - 1
                    area: int = height * width
                    if area > best:
                        best = area
                top += 1
                stack[top] = j
            return best

        for i in range(rows):
            row = matrix[i]
            # 高さ更新
            for j in range(cols):
                heights[j] = heights[j] + 1 if row[j] == "1" else 0
            # 行を底にした最大長方形
            area = largest_rectangle_in_histogram(heights)
            if area > max_area:
                max_area = area

        return max_area

    # 参考: 業務開発向けの堅牢版（LeetCodeでは呼ばれない）
    def maximalRectangle_production(self, matrix: List[List[str]]) -> int:
        """
        業務開発向け: 型/範囲検証を行い、仕様違反時は例外を送出。

        Raises:
            TypeError: 要素型が '0'/'1' 以外
            ValueError: 行長不揃い、サイズ範囲外など
        """
        if not isinstance(matrix, list):
            raise TypeError("matrix must be a list of lists")
        if not matrix:
            return 0
        cols: int = len(matrix[0])
        if cols == 0:
            return 0
        rows: int = len(matrix)
        if not (1 <= rows <= 200):
            raise ValueError("rows must be within [1, 200]")
        if not (1 <= cols <= 200):
            raise ValueError("cols must be within [1, 200]")
        for r, row in enumerate(matrix):
            if not isinstance(row, list) or len(row) != cols:
                raise ValueError("All rows must have identical length")
            for c, v in enumerate(row):
                if not isinstance(v, str) or (v != "0" and v != "1"):
                    raise TypeError("matrix[i][j] must be '0' or '1'")

        # 入力が妥当なら競技向け実装に委譲（コピーせず同じ参照を渡す）
        return self.maximalRectangle(matrix)


# ```

# ---

# ### 5. 検証（方針のみ）

# * **境界値**: `[]` → 0、`[["0"]]` → 0、`[["1"]]` → 1、全1の行列、全0の行列、縦長/横長ケース、ジグザグ。
# * **上限**: 200×200 で例外なく O(R·C) で処理可能。
# * **型チェック**: `maximalRectangle_production` による静的/動的検証で *pylance* も問題なし。

File: DataStructures/Stacks/leetcode/85. Maximal Rectangle/GPT/Maximal_Rectangle.ts
// ### 1. 問題の分析

// * **競技プログラミング視点での分析**

//   * 各行を底とする「連続1の高さ」をヒストグラムとして更新し、行ごとに「単調増加スタック」で最大長方形を計算すると **O(R×C)**。
//   * 追加メモリは `heights`（列数）とスタック（列数）で **O(C)**。配列を再利用し、push/pop を避けて「手動スタック（添字管理）」にすると高速・省メモリ。

// * **業務開発視点での分析**

//   * 型は `string[][]`（'0'/'1'）を厳密に扱い、行・列の不揃いを実行前に検証。
//   * 実装は純粋関数化し、副作用ゼロ。ローカル関数としてヒストグラム計算を分離して見通しを確保。
//   * 仕様違反入力には `TypeError` / `RangeError` を早期に投げる（ホットパス外）。

// * **TypeScript特有の考慮点**

//   * `readonly` パラメータを受け取りつつ内部は数値配列に正規化。
//   * 関数境界の型で不正入力を事前に抑止し、実行時ガードは最小限。
//   * 配列は number 単型を維持し、hidden class/最適化を阻害しない単純な for ループで実装。

// ---

// ### 2. アルゴリズムアプローチ比較

// | アプローチ                          | 時間計算量           |  空間計算量 | TS実装コスト | 型安全性 | 可読性 | 備考           |
// | ------------------------------ | --------------- | -----: | ------: | ---: | --: | ------------ |
// | 方法A: 行ごとのヒストグラム + 単調増加スタック（採用） | O(R·C)          |   O(C) |       低 |    高 |   高 | 最速定番、実装短・バグ少 |
// | 方法B: DPで各セルの左幅/右幅/高さを保持        | O(R·C)          | O(R·C) |       中 |    高 |   中 | メモリ重め、コード量増  |
// | 方法C: 各セルから拡張して全探索              | O(R·C·min(R,C)) |   O(1) |       低 |    高 |   中 | 小規模のみ現実的     |

// ---

// ### 3. 選択したアルゴリズムと理由

// * **選択したアプローチ**: 方法A（行ごとのヒストグラム + 単調増加スタック）
// * **理由**:

//   * **計算量的な優位性**: 制約最大でも O(R·C) で余裕、メモリ O(C)。
//   * **TypeScript環境での型安全性**: `string` → `number`（高さ）を限定変換し、内部は number 単型で最適化。
//   * **保守性・可読性**: 主処理は 2 つのループ + 小さなローカル関数。副作用なしで検証容易。
// * **TypeScript特有の最適化ポイント**:

//   * `readonly` 入力を受けつつ、内部配列は再利用（`heights`/`stack` の固定長・手動トップ管理）。
//   * 例外はホットパス外の先頭チェックのみ。
//   * 追加のオブジェクト生成を避け、単純 for で JIT に優しい形。

// ---

// ### 4. 実装コード（LeetCode形式 / Node.js v18, **ESM** / 外部ライブラリ不可）

// ```typescript
/* eslint-disable @typescript-eslint/explicit-function-return-type */
/* eslint-disable @typescript-eslint/no-unused-vars */
// ESM 形式（ローカル実行時は `node --loader ts-node/esm solution.ts` 等の想定）
// LeetCode 環境では関数定義のみで可。ここでは ESM としてエクスポートも付与。

/**
 * 最大長方形（LeetCode 85）
 * 与えられた '0'/'1' 文字行列で、1のみから成る最大長方形の面積を返す。
 *
 * @param matrix - R×C の2次元配列（各要素は '0' または '1'）
 * @returns 最大長方形の面積
 * @throws {TypeError} matrix が 2 次元配列でない／要素が '0'/'1' 以外
 * @throws {RangeError} 行数/列数が [1,200] を外れる、または列長が不揃い
 * @complexity Time: O(R*C), Space: O(C)
 */
export default function maximalRectangle(matrix: readonly (readonly string[])[]): number {
    // ---- 入力検証（軽量＆早期；ホットパス外） ----
    if (!Array.isArray(matrix)) {
        throw new TypeError('matrix must be a 2D array');
    }
    const rows = matrix.length;
    if (!(rows >= 1 && rows <= 200)) {
        throw new RangeError('rows must be within [1, 200]');
    }
    if (!Array.isArray(matrix[0])) {
        throw new TypeError('matrix must be a 2D array (array of arrays)');
    }
    const cols = matrix[0].length;
    if (!(cols >= 1 && cols <= 200)) {
        throw new RangeError('cols must be within [1, 200]');
    }
    for (let i = 0; i < rows; i++) {
        const row = matrix[i];
        if (!Array.isArray(row) || row.length !== cols) {
            throw new RangeError('All rows must be arrays of identical length');
        }
        for (let j = 0; j < cols; j++) {
            const v = row[j];
            if (typeof v !== 'string' || (v !== '0' && v !== '1')) {
                throw new TypeError("matrix[i][j] must be '0' or '1'");
            }
        }
    }

    // ---- 本処理：行ごとのヒストグラム + 単調増加スタック ----
    // heights[j]: 現在行を底とする列 j の連続 1 の高さ
    const heights: number[] = new Array<number>(cols).fill(0);
    // スタックは「インデックス」を格納。push/pop を避け、top を手動管理で高速化。
    const stack: number[] = new Array<number>(cols + 1);
    let maxArea = 0;

    // ヒストグラムの最大長方形を O(C) で計算（センチネルはループ境界で擬似化）
    function largestRectangleInHistogram(h: readonly number[]): number {
        let best = 0;
        let top = -1; // スタックのトップ（-1 は空）

        for (let j = 0; j <= cols; j++) {
            const cur = j === cols ? 0 : h[j];
            while (top >= 0 && cur < h[stack[top]]) {
                const height = h[stack[top--]];
                const leftLess = top >= 0 ? stack[top] : -1;
                const width = j - leftLess - 1;
                const area = height * width;
                if (area > best) best = area;
            }
            stack[++top] = j;
        }
        return best;
    }

    for (let i = 0; i < rows; i++) {
        const row = matrix[i];
        for (let j = 0; j < cols; j++) {
            // '1' なら高さをインクリメント、'0' ならリセット
            heights[j] = row[j] === '1' ? heights[j] + 1 : 0;
        }
        const area = largestRectangleInHistogram(heights);
        if (area > maxArea) maxArea = area;
    }

    return maxArea;
}

// LeetCode 用の余計なエクスポート抑止（環境により不要）
// export {};
// ```

// ---

// ## TypeScript固有の最適化観点

// * **型安全性の活用**

//   * `readonly (readonly string[])[]` で外部からの破壊を禁止、実装側では number 配列に正規化して単型を維持。
//   * 実行時ガードで仕様外入力を早期排除（`TypeError` / `RangeError`）。
// * **コンパイル時最適化**

//   * 単純な for ループと number 単型配列で JIT の最適化を阻害しない。
//   * スタックは配列再利用 + 手動 `top` 管理で割当・GC を削減。
// * **開発効率と保守性**

//   * 主処理は 2 段（高さ更新／ヒストグラム計算）に明確分割、ローカル関数でカプセル化。
//   * 例外はホットパス外に集約し、データサイズ上限の逸脱も検知。
